crates/rome_js_analyze/src/analyzers/nursery/no_non_null_assertion.rs

```diff
@@ -0,0 +1,178 @@
+use crate::JsRuleAction;
+use rome_analyze::{context::RuleContext, declare_rule, ActionCategory, Ast, Rule, RuleDiagnostic};
+use rome_console::markup;
+use rome_diagnostics::Applicability;
+use rome_js_factory::make;
+use rome_js_syntax::{AnyJsExpression, JsSyntaxKind, TsNonNullAssertionExpression};
+use rome_rowan::{AstNode, BatchMutationExt, SyntaxNodeCast};
+
+declare_rule! {
+    /// Disallow non-null assertions using the `!` postfix operator.
+    ///
+    /// TypeScript's `!` non-null assertion operator asserts to the type system that an expression is non-nullable, as 
+    /// in not `null` or `undefined`. Using assertions to tell the type system new information is often a sign that 
+    /// code is not fully type-safe. It's generally better to structure program logic so that TypeScript understands 
+    /// when values may be nullable.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```ts,expect_diagnostic
+    /// interface Example {
+    ///   property?: string;
+    /// }
+    /// declare const example: Example;
+    /// const includesBaz = foo.property!.includes('baz');
+    /// ```
+    /// 
+    /// ### Valid
+    /// 
+    /// ```ts
+    /// interface Example {
+    ///   property?: string;
+    /// }
+    /// 
+    /// declare const example: Example;
+    /// const includesBaz = foo.property?.includes('baz') ?? false;
+    /// ```
+    ///
+    pub(crate) NoNonNullAssertion {
+        version: "11.0.0",
+        name: "noNonNullAssertion",
+        recommended: false,
+    }
+}
+
+fn can_replace_with_optional_chain(node: &TsNonNullAssertionExpression) -> bool {
+    match node.parent::<AnyJsExpression>() {
+        Some(parent) => {
+            matches!(
+                parent.syntax().kind(),
+                JsSyntaxKind::JS_STATIC_MEMBER_EXPRESSION
+                    | JsSyntaxKind::JS_COMPUTED_MEMBER_EXPRESSION
+                    | JsSyntaxKind::JS_CALL_EXPRESSION
+            )
+        }
+        None => false,
+    }
+}
+
+impl Rule for NoNonNullAssertion {
+    type Query = Ast<TsNonNullAssertionExpression>;
+    type State = (TsNonNullAssertionExpression, bool);
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        if let Some(child) = node.syntax().first_child() {
+            let is_child_non_null_assertion =
+                matches!(child.kind(), JsSyntaxKind::TS_NON_NULL_ASSERTION_EXPRESSION);
+            if is_child_non_null_assertion {
+                return None;
+            }
```

Comment:
It's better to use the typed accessors on the node to get additional type safety.
```suggestion
		let inner = node.expression().ok()?; 
		
		match inner {
			AnyJsExpression::TsNonNullAssertionExpression(_) => None,
			inner => {
			  ..
			}
```